### DISCUSSION TITLE
query_range result is not accurate when step is too large

### DISCUSSION THREAD

**Discussion by TIBCOtzhu** [prometheus/prometheus#11820] [Q&A]
### What did you do?

Update metric "input_record_count", then uses /query_range api to query it. 

query=input_record_count
start=2023-01-01T00:00:00.000Z
end=2023-01-07T00:00:00.000Z

when the step=2h, the last metric value of container "flow1-fhqyswby" is 51, timestamp is 1673042400

when the step=12h, the last metric value of container "flow1-fhqyswby" is 44, timestamp is 1673006400

I am wondering what causes the last metric values are different.

Thanks.

### What did you see instead? Under which circumstances?

http://localhost:53718/api/v1/query_range?query=input_record_count&start=2023-01-01T00:00:00.000Z&end=2023-01-07T00:00:00.000Z&step=12h

> {
>    "status":"success",
>    "data":{
>       "resultType":"matrix",
>       "result":[
>          {
>             "metric":{
>                "__name__":"input_record_count",
>                "container":"flow1-fhqyswby",
>                "endpoint":"admin-ui",
>                "environment":"env1",
>                "flowName":"Flow_1",
>                "instance":"10.224.1.255:8008",
>                "namespace":"datachannels",
>                "pod":"flow1-fhqyswby-0",
>                "username":"admin"
>             },
>             "values":[
>                [
>                   1672963200,
>                   "37"
>                ],
>                [
>                   1673006400,
>                   "44"
>                ]
>             ]
>          },
>          {
>             "metric":{
>                "__name__":"input_record_count",
>                "container":"flow1-nxlicnjl",
>                "endpoint":"admin-ui",
>                "environment":"Development",
>                "flowName":"Flow_1",
>                "instance":"10.224.1.186:8008",
>                "namespace":"development",
>                "pod":"flow1-nxlicnjl-0",
>                "username":"admin"
>             },
>             "values":[
>                [
>                   1672531200,
>                   "2"
>                ],
>                [
>                   1672574400,
>                   "2"
>                ],
>                [
>                   1672617600,
>                   "2"
>                ],
>                [
>                   1672660800,
>                   "2"
>                ],
>                [
>                   1672704000,
>                   "2"
>                ],
>                [
>                   1672747200,
>                   "2"
>                ],
>                [
>                   1672790400,
>                   "2"
>                ],
>                [
>                   1672833600,
>                   "2"
>                ],
>                [
>                   1672876800,
>                   "2"
>                ],
>                [
>                   1672920000,
>                   "2"
>                ],
>                [
>                   1672963200,
>                   "2"
>                ],
>                [
>                   1673006400,
>                   "2"
>                ]
>             ]
>          }
>       ]
>    }
> }

http://localhost:53718/api/v1/query_range?query=input_record_count&start=2023-01-01T00:00:00.000Z&end=2023-01-07T00:00:00.000Z&step=2h

> {
>    "status":"success",
>    "data":{
>       "resultType":"matrix",
>       "result":[
>          {
>             "metric":{
>                "__name__":"input_record_count",
>                "container":"flow1-fhqyswby",
>                "endpoint":"admin-ui",
>                "environment":"Data-Channels",
>                "flowName":"Flow_1",
>                "instance":"10.224.1.255:8008",
>                "namespace":"datachannels",
>                "pod":"flow1-fhqyswby-0",
>                "username":"admin"
>             },
>             "values":[
>                [
>                   1672948800,
>                   "34"
>                ],
>                [
>                   1672956000,
>                   "37"
>                ],
>                [
>                   1672963200,
>                   "37"
>                ],
>                [
>                   1672970400,
>                   "44"
>                ],
>                [
>                   1672977600,
>                   "44"
>                ],
>                [
>                   1672984800,
>                   "44"
>                ],
>                [
>                   1672992000,
>                   "44"
>                ],
>                [
>                   1672999200,
>                   "44"
>                ],
>                [
>                   1673006400,
>                   "44"
>                ],
>                [
>                   1673013600,
>                   "44"
>                ],
>                [
>                   1673020800,
>                   "51"
>                ],
>                [
>                   1673028000,
>                   "51"
>                ],
>                [
>                   1673035200,
>                   "51"
>                ],
>                [
>                   1673042400,
>                   "51"
>                ]
>             ]
>          },
>          {
>             "metric":{
>                "__name__":"input_record_count",
>                "container":"flow1-nxlicnjl",
>                "endpoint":"admin-ui",
>                "environment":"Development",
>                "flowName":"Flow_1",
>                "instance":"10.224.1.186:8008",
>                "namespace":"development",
>                "pod":"flow1-nxlicnjl-0",
>                "username":"admin"
>             },
>             "values":[
>                [
>                   1672531200,
>                   "2"
>                ],
>                [
>                   1672538400,
>                   "2"
>                ],
>                [
>                   1672545600,
>                   "2"
>                ],
>                [
>                   1672552800,
>                   "2"
>                ],
>                [
>                   1672560000,
>                   "2"
>                ],
>                [
>                   1672567200,
>                   "2"
>                ],
>                [
>                   1672574400,
>                   "2"
>                ],
>                [
>                   1672581600,
>                   "2"
>                ],
>                [
>                   1672588800,
>                   "2"
>                ],
>                [
>                   1672596000,
>                   "2"
>                ],
>                [
>                   1672603200,
>                   "2"
>                ],
>                [
>                   1672610400,
>                   "2"
>                ],
>                [
>                   1672617600,
>                   "2"
>                ],
>                [
>                   1672624800,
>                   "2"
>                ],
>                [
>                   1672632000,
>                   "2"
>                ],
>                [
>                   1672639200,
>                   "2"
>                ],
>                [
>                   1672646400,
>                   "2"
>                ],
>                [
>                   1672653600,
>                   "2"
>                ],
>                [
>                   1672660800,
>                   "2"
>                ],
>                [
>                   1672668000,
>                   "2"
>                ],
>                [
>                   1672675200,
>                   "2"
>                ],
>                [
>                   1672682400,
>                   "2"
>                ],
>                [
>                   1672689600,
>                   "2"
>                ],
>                [
>                   1672696800,
>                   "2"
>                ],
>                [
>                   1672704000,
>                   "2"
>                ],
>                [
>                   1672711200,
>                   "2"
>                ],
>                [
>                   1672718400,
>                   "2"
>                ],
>                [
>                   1672725600,
>                   "2"
>                ],
>                [
>                   1672732800,
>                   "2"
>                ],
>                [
>                   1672740000,
>                   "2"
>                ],
>                [
>                   1672747200,
>                   "2"
>                ],
>                [
>                   1672754400,
>                   "2"
>                ],
>                [
>                   1672761600,
>                   "2"
>                ],
>                [
>                   1672768800,
>                   "2"
>                ],
>                [
>                   1672776000,
>                   "2"
>                ],
>                [
>                   1672783200,
>                   "2"
>                ],
>                [
>                   1672790400,
>                   "2"
>                ],
>                [
>                   1672797600,
>                   "2"
>                ],
>                [
>                   1672804800,
>                   "2"
>                ],
>                [
>                   1672812000,
>                   "2"
>                ],
>                [
>                   1672819200,
>                   "2"
>                ],
>                [
>                   1672826400,
>                   "2"
>                ],
>                [
>                   1672833600,
>                   "2"
>                ],
>                [
>                   1672840800,
>                   "2"
>                ],
>                [
>                   1672848000,
>                   "2"
>                ],
>                [
>                   1672855200,
>                   "2"
>                ],
>                [
>                   1672862400,
>                   "2"
>                ],
>                [
>                   1672869600,
>                   "2"
>                ],
>                [
>                   1672876800,
>                   "2"
>                ],
>                [
>                   1672884000,
>                   "2"
>                ],
>                [
>                   1672891200,
>                   "2"
>                ],
>                [
>                   1672898400,
>                   "2"
>                ],
>                [
>                   1672905600,
>                   "2"
>                ],
>                [
>                   1672912800,
>                   "2"
>                ],
>                [
>                   1672920000,
>                   "2"
>                ],
>                [
>                   1672927200,
>                   "2"
>                ],
>                [
>                   1672934400,
>                   "2"
>                ],
>                [
>                   1672941600,
>                   "2"
>                ],
>                [
>                   1672948800,
>                   "2"
>                ],
>                [
>                   1672956000,
>                   "2"
>                ],
>                [
>                   1672963200,
>                   "2"
>                ],
>                [
>                   1672970400,
>                   "2"
>                ],
>                [
>                   1672977600,
>                   "2"
>                ],
>                [
>                   1672984800,
>                   "2"
>                ],
>                [
>                   1672992000,
>                   "2"
>                ],
>                [
>                   1672999200,
>                   "2"
>                ],
>                [
>                   1673006400,
>                   "2"
>                ],
>                [
>                   1673013600,
>                   "2"
>                ],
>                [
>                   1673020800,
>                   "2"
>                ],
>                [
>                   1673028000,
>                   "2"
>                ],
>                [
>                   1673035200,
>                   "2"
>                ],
>                [
>                   1673042400,
>                   "2"
>                ]
>             ]
>          }
>       ]
>    }
> }

### System information

Linux 5.4.0-1094-azure x86_64

### Prometheus version

```text
prometheus, version 2.32.1 (branch: HEAD, revision: 41f1a8125e664985dd30674e5bdf6b683eff5d32)
  build user:       root@54b6dbd48b97
  build date:       20211217-22:08:06
  go version:       go1.17.5
  platform:         linux/amd64
```

**Comment by roidelapluie** [comment]
The difference in the last metric value for the same container and time period between the two queries is due to the different step sizes used in the queries, and it is expected, not a bug.

The step parameter in a Prometheus query determines the resolution of the returned data. A smaller step size will result in a larger amount of data points being returned, each with a smaller time interval between them. A larger step size will result in fewer data points, each with a larger time interval between them.

In the first query, with a step size of 12 hours, the last metric value for the container "flow1-fhqyswby" is 44 at a timestamp of 1673006400. In the second query, with a step size of 2 hours, the last metric value for the same container and time period is 51 at a timestamp of 1673042400. This means that there were more input records counted for the container within the 2 hour interval leading up to 1673042400 compared to the 12 hour interval leading up to 1673006400.

**1673042400 is 10 hours after 1673006400.**

I hope this helps to explain the difference in the last metric values between the two queries. Let me know if you have any further questions.

**Reply by TIBCOtzhu** [reply]
Thank you for your answer.

Just a further question for 12 hours interval, since the start time is 2023-01-01, 00:00:00, and the end time is 2023-01-07-00:00:00, why the last timestamp is 1673006400 (Friday, January 6, 2023 12:00:00), should it be 1673049599(Friday, January 6, 2023 23:59:59)? Does it mean if the total time between the last timestamp and the end time is smaller than the step, then the data within this time range won't be reflected in the result?

**Reply by roidelapluie** [reply]
When using the /query_range API with a step parameter of 12h, Prometheus will return the data in 12 hour intervals, starting from the timestamp specified in the start parameter and ending at the timestamp specified in the end parameter. The last timestamp returned in the query will be the last complete interval that falls within the queried range, in this case, the last complete 12 hour interval ending at 1673006400 (Friday, January 6, 2023 12:00:00).
